Tests/BackendIntegrationTests/SubscriberAttributesManagerIntegrationTests.swift

```diff
@@ -0,0 +1,157 @@
+
+        Purchases.shared.setEmail(email)
+        errors = await self.syncAttributes()
+        verifyAttributesSyncedWithNoErrors(errors, 0)
+    }
+
+    func testChangingEmailSyncsIt() async throws {
+        Purchases.shared.setEmail("test@revenuecat.com")
+
+        var errors = await self.syncAttributes()
+        verifyAttributesSyncedWithNoErrors(errors, 1)
+
+        Purchases.shared.setEmail("test2@revenuecat.com")
+        errors = await self.syncAttributes()
+        verifyAttributesSyncedWithNoErrors(errors, 1)
+    }
+
+    func testSyncInvalidEmail() async throws {
+        Purchases.shared.setEmail("invalid @ email @.com")
+
+        let errors = await self.syncAttributes()
+        expect(errors).to(haveCount(1))
+
+        let error = try XCTUnwrap(errors.first ?? nil) as NSError
+        expect(error.domain) == RCPurchasesErrorCodeDomain
+        expect(error.code) == ErrorCode.invalidSubscriberAttributesError.rawValue
+        expect(error.subscriberAttributesErrors) == [
+            "$email": "Email address is not a valid email."
+        ]
+    }
+
+    func testLogInGetsNewAttributes() async throws {
+        Purchases.shared.setEmail("test@revenuecat.com")
+
+        var errors = await self.syncAttributes()
+        verifyAttributesSyncedWithNoErrors(errors, 1)
+
+        _ = try await Purchases.shared.logIn(UUID().uuidString)
+
+        Purchases.shared.setEmail("test@revenuecat.com")
+
+        errors = await self.syncAttributes()
+        verifyAttributesSyncedWithNoErrors(errors, 1)
+    }
+
+    func testPushTokenWithInvalidTokenDoesNotFail() async throws {
+        Purchases.shared.setPushToken("invalid token".data(using: .utf8))
+
+        let errors = await self.syncAttributes()
+        verifyAttributesSyncedWithNoErrors(errors, 1)
+    }
+
+    func testSetCustomAttributes() async throws {
+        Purchases.shared.setAttributes([
+            "custom_key": "random value",
+            "locale": Locale.current.identifier
+        ])
+
+        let errors = await self.syncAttributes()
+        verifyAttributesSyncedWithNoErrors(errors, 1) // 1 user with 2 attributes
+    }
+
+    func testSetMultipleAttributes() async throws {
+        Purchases.shared.setDisplayName("Tom Hanks")
+        Purchases.shared.setPhoneNumber("4157689215")
+
+        let errors = await self.syncAttributes()
+        verifyAttributesSyncedWithNoErrors(errors, 1) // 1 user with 2 attributes
+    }
+
+    func testSetAttributesForMultipleUsers() async throws {
+        _ = try await Purchases.shared.logIn(UUID().uuidString)
+        Purchases.shared.setDisplayName("User 1")
+
+        _ = try await Purchases.shared.logIn(UUID().uuidString)
+        Purchases.shared.setDisplayName("User 2")
+
+        let errors = await self.syncAttributes()
+        verifyAttributesSyncedWithNoErrors(errors, 2)
+    }
+
+}
+
+private extension SubscriberAttributesManagerIntegrationTests {
+
+    func syncAttributes() async -> [Error?] {
```

Comment:
`async` makes these tests so much better.

Reply:
Yeah, reading and making sense of the tests is so easy with out needing to build special block handling/waiting